### PR TITLE
support more image dtypes

### LINKED ISSUE #447: Some types (int64) are not supported yet.
## 🐛 Bug

By mistake i tried to add an image (array) of type int64 and boom -- it crashes.

We might want to have some tests or examples for more exotic types.
For example, complex types would be cool but are outside the scope of this issue...
(but exciting!)

## To Reproduce

Create an array of int64 dtype and try to view it in napari... 

You get this:

```
WARNING: Traceback (most recent call last):
  File "/Users/royer/workspace/python/pitl/pitl/it/demo/demo_it_n2s_gbm_2D_rgb.py", line 76, in <module>
    demo(astronaut())
  File "/Users/royer/workspace/python/pitl/pitl/it/demo/demo_it_n2s_gbm_2D_rgb.py", line 31, in demo
    rescale_intensity(noisy, in_range='image', out_range=(0, 1)), name='noisy'
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/napari/_qt/event_loop.py", line 31, in gui_qt
    app.exec_()
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 501, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 818, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/scene/canvas.py", line 217, in on_draw
    self._draw_scene()
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/scene/canvas.py", line 266, in _draw_scene
    self.draw_visual(self.scene)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/scene/canvas.py", line 304, in draw_visual
    node.draw()
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/scene/visuals.py", line 99, in draw
    self._visual_superclass.draw(self)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/visuals/visual.py", line 443, in draw
    self._vshare.index_buffer)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/visuals/shaders/program.py", line 101, in draw
    Program.draw(self, *args, **kwargs)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/gloo/program.py", line 533, in draw
    canvas.context.flush_commands()
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/gloo/context.py", line 176, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/gloo/glir.py", line 572, in flush
    self._shared.flush(parser)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/gloo/glir.py", line 494, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/gloo/glir.py", line 819, in parse
    self._parse(command)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/gloo/glir.py", line 787, in _parse
    ob.set_data(*args)
  File "/Users/royer/anaconda3/envs/pitl/lib/python3.7/site-packages/vispy/gloo/glir.py", line 1537, in set_data
    raise ValueError("Type %r not allowed for texture" % data.dtype)
ValueError: Type dtype('int64') not allowed for texture
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PySide2) at 0x128724438>> for DrawEvent
WARNING: Error drawing visual <Image at 0x133fde5f8>
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PySide2) at 0x128724438>> repeat 2
```

## Environment

 - napari Version (e.g., 1.0): 0.1.3
 - OS (e.g., Linux): OSX
 - Python version: 3.7
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR fixes #447 by extending support to all common dtpyes using the method suggested by @kne42. I have verified that there was a real error caused by #447 and that these changes fix it, and I have added tests that should catch the previous error - however these tests also passed before making the changes in this PR implying that the code that is causing the error in #447 was not being run in our tests. I'm not sure why that is the case, it's maybe because of multithreading issues, like the vispy draw is being run on another thread that can die without causing the test to fail? I'm not sure, i tried adding in all sorts of `canvas.update` calls etc to try and make sure the code was being run, but i couldn't get the tests to fail (though if i tried running similar code in a script it would fail). Any ideas on the cause of this @kne42? We should try and figure out before we merge. 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [x] add `tests/test_dtypes.py`

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/_vispy/vispy_image_layer.py [hunk 1]
```diff
@@ -43,11 +52,22 @@ def _on_display_change(self):
         self.reset()
 
     def _on_data_change(self):
+        data = self.layer._data_view
+        dtype = np.dtype(data.dtype)
+        if dtype not in texture_dtypes:
+            try:
+                dtype = dict(i=np.int16, f=np.float64)[dtype.kind]
```
Comment (jni) [inline napari/_vispy/vispy_image_layer.py]: Can you explain this line?

**Review by jni — APPROVED** [review]
@sofroniewn there's a line that needs explanation for me, and it would definitely be great to get to the bottom of the testing problem, but otherwise this LGTM, thank you!

File: napari/_vispy/vispy_image_layer.py (hunk 1 above)
Comment (kne42) [reply napari/_vispy/vispy_image_layer.py]: @jni this is equivalent to
```python
if dtype.kind == 'i':    # is an int
    dtype = np.int16
elif dtype.kind == 'f':  # is a float
    dtype = np.float64
else:
    raise TypeError(f'type {dtype} not allowed for texture; must be one of {set(texture_dtypes)}')
```

Comment (jni) [reply napari/_vispy/vispy_image_layer.py]: Ok, but why is np.int6 the default int type?

Comment (kne42) [reply napari/_vispy/vispy_image_layer.py]: @sofroniewn ^

Comment (kne42) [reply napari/_vispy/vispy_image_layer.py]: also see [here](https://github.com/vispy/vispy/blob/maint/0.6/vispy/gloo/glir.py#L1409): `np.float64` is not a valid option at this time

would recommend you directly set `texture_dtypes = vispy.gloo.glir.GlirTexture._types.keys()` [as suggested in my post](https://github.com/napari/napari/issues/447#issuecomment-522079016)

bonus points if you don't hardcode the default and figure them out based on `texture_dtypes` ;)

**Review by kne42 — DISMISSED** [review]
see [this thread reply](https://github.com/napari/napari/pull/498/files/2f2a1f36b07990bfb464a6647728e7a21a4f7e97#r319568316)

File: napari/_vispy/vispy_image_layer.py (hunk 1 above)
Comment (sofroniewn) [reply napari/_vispy/vispy_image_layer.py]: I started with your recommendation, but found that list diverged from what threw errors when I tried running things. `np.float64` ran fine, but wasn't on that list. `np.int32` failed despite being on that list, which is what lead me to the hard coding approach that I took. maybe i got something wrong when trying them, but that was my experience

**Comment by sofroniewn** [comment]
@kne42 do you have any idea why the tests weren't catching these errors?

File: napari/_vispy/vispy_image_layer.py (hunk 1 above)
Comment (royerloic) [reply napari/_vispy/vispy_image_layer.py]: Just a note: I would avoid anything else than uint8, uint16, float32 for vispy/OpenGL textures. Support for the other types such as 16bit and 64bit types might be spotty or not fully supported in all machines. Also, 64 bit textures is overkill and a waste of precious GPU memory, so I would avoid that anyway -- particularly for 3D textures (volume rendering).

Comment (sofroniewn) [reply napari/_vispy/vispy_image_layer.py]: ok now `float64` will be converted to `float32` before passing to vispy

**Comment by sofroniewn** [comment]
I've checked that this works now and made the changes requested by @royerloic and am going to merge after the tests pass.

We still need to figure out why the errors in our vispy views aren't being caught by our tests, but that can be figured out separately

**Pull request merged**